### PR TITLE
Create script to deploy stable Eclipse Che in Openshift CI and check if it is possible to update it to nightly

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: Flavius Lacatusu <flacatus@redhat.com>

N/A
### What does this PR do?
Create script to deploy stable Eclipse Che in Openshift CI and check if it is possible to update it to nightly

### Screenshot/screencast of this PR

N/A
### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/19803

### How to test this PR?

N/A

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [Nightly OLM bundle is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-nightly-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: .ci/oci-nightly-update.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+export OPERATOR_REPO=$(dirname $(dirname $(readlink -f "$0")));
+source "${OPERATOR_REPO}"/.github/bin/common.sh
+source "${OPERATOR_REPO}"/.github/bin/oauth-provision.sh
+
+#Stop execution on any error
+trap "catchFinish" EXIT SIGINT
+
+overrideDefaults() {
+  # CI_CHE_OPERATOR_IMAGE it is che operator image builded in openshift CI job workflow. More info about how works image dependencies in ci:https://github.com/openshift/ci-tools/blob/master/TEMPLATES.md#parameters-available-to-templates
+  export OPERATOR_IMAGE=${CI_CHE_OPERATOR_IMAGE:-"quay.io/eclipse/che-operator:nightly"}
```
Comment (tolusha) [inline .ci/oci-nightly-update.sh:32]: Let's not to set a default image. It is better to fail.

File: .ci/oci-nightly-update.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+runTests() {
+  deployEclipseCheStableWithChectl ${LAST_PACKAGE_VERSION}
+  provisionOAuth
+  createWorkspace
+
+  # Update Eclipse Che to nightly and start workspace
+  initLatestTemplates
+  chectl server:update --yes --templates="${TEMPLATES}"
+  patchEclipseCheOperatorSubscription
```
Comment (tolusha) [inline .ci/oci-nightly-update.sh:43]: Do we need to patch subscription if `--installer operator`  was used?

File: .ci/oci-nightly-update.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  export OPERATOR_IMAGE=${CI_CHE_OPERATOR_IMAGE:-"quay.io/eclipse/che-operator:nightly"}
+}
+
+runTests() {
+  deployEclipseCheStableWithChectl ${LAST_PACKAGE_VERSION}
+  provisionOAuth
+  createWorkspace
+
+  # Update Eclipse Che to nightly and start workspace
+  initLatestTemplates
```
Comment (tolusha) [inline .ci/oci-nightly-update.sh:41]: Can we move this before running tests?

File: .github/bin/common.sh [hunk 4]
```diff
@@ -69,6 +69,10 @@ initStableTemplates() {
   export LAST_PACKAGE_VERSION
   PREVIOUS_PACKAGE_VERSION=$(echo "${versions[@]}" | jq '.data.repository.refs.edges[0].node.name | sub("\""; "")' | tr -d '"')
   export PREVIOUS_PACKAGE_VERSION
+}
+
+initStableTemplates() {
```
Comment (tolusha) [inline .github/bin/common.sh:74]: `initStableTemplates` is used by other scripts.
I am afraid if `LAST_PACKAGE_VERSION` and `PREVIOUS_PACKAGE_VERSION` will be undefined there.
Can we revert this and refactor next time?

File: .github/bin/common.sh [hunk 5]
```diff
@@ -406,3 +410,13 @@ waitWorkspaceStartedDevWorkspaceController() {
 createWorkspaceDevWorkspaceCheOperator() {
   oc apply -f https://raw.githubusercontent.com/che-incubator/devworkspace-che-operator/main/samples/flattened_theia-nodejs.yaml -n ${NAMESPACE}
 }
+
+deployEclipseCheStableWithChectl() {
```
Comment (tolusha) [inline .github/bin/common.sh:414]: I would like to see something like `deployEclipseChe` with passing arguments to the function to make it reusable.

File: .github/bin/common.sh (hunk 4 above)
Comment (flacatus) [reply .github/bin/common.sh:74]: I'm not renaming this function. The same function still exists. I'm just creating a new function which get the Che versions from GH.

File: .ci/oci-nightly-update.sh (hunk 2 above)
Comment (flacatus) [reply .ci/oci-nightly-update.sh]: I renamed the function because `patchEclipseCheOperatorSubscription` is just patching the che operator image

**Review by tolusha — APPROVED** [review]

**Pull request merged**